cardano-crypto-class/src/Cardano/Crypto/DSIGN/Ed25519.hs

```diff
@@ -76,37 +121,60 @@ instance DSIGNAlgorithm Ed25519DSIGN where
     type Signable Ed25519DSIGN = SignableRepresentation
 
     signDSIGN () a (SignKeyEd25519DSIGN sk) =
-        let vk = toPublic sk
-            bs = getSignableRepresentation a
-         in SigEd25519DSIGN $ sign sk vk bs
+      let bs = getSignableRepresentation a
+      in SigEd25519DSIGN $ unsafeDupablePerformIO $
+            BS.useAsCStringLen bs $ \(ptr, len) -> 
+            psbUseAsSizedPtr sk $ \skPtr ->
+            allocaSized $ \pkPtr -> do
+                cOrError "signDSIGN @Ed25519DSIGN" "c_crypto_sign_ed25519_sk_to_pk"
+                  $ c_crypto_sign_ed25519_sk_to_pk pkPtr skPtr
+                psbCreateSized $ \sigPtr -> do
+                  cOrError "signDSIGN @Ed25519DSIGN" "c_crypto_sign_ed25519_detached"
+                    $ c_crypto_sign_ed25519_detached sigPtr nullPtr (castPtr ptr) (fromIntegral len) skPtr
 
     verifyDSIGN () (VerKeyEd25519DSIGN vk) a (SigEd25519DSIGN sig) =
-        if verify vk (getSignableRepresentation a) sig
-          then Right ()
-          else Left "Verification failed"
+        let bs = getSignableRepresentation a
+        in unsafeDupablePerformIO $
+          BS.useAsCStringLen bs $ \(ptr, len) ->
+          psbUseAsSizedPtr vk $ \vkPtr ->
+          psbUseAsSizedPtr sig $ \sigPtr -> do
+              res <- c_crypto_sign_ed25519_verify_detached sigPtr (castPtr ptr) (fromIntegral len) vkPtr
+              if res == 0
+              then return (Right ())
+              else do
+                  -- errno <- getErrno
+                  return (Left  "Verification failed")
```

Comment:
This may actually be the result of copy-pasta. I don't think we necessarily *need* the errno, but OTOH it won't hurt to include it, unless something somewhere depends on the exact error message. I'd be fine with it either way.